### PR TITLE
Fix parsing of SETTINGS clause of the INSERT ... SELECT ... SETTINGS query

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix parsing of SETTINGS clause of the INSERT ... SELECT ... SETTINGS query

Detailed description / Documentation draft:
Before this patch the following query ignores the settings for INSERT:
```sql
   insert into test_parallel_insert select * from numbers_mt(65535*2) settings max_insert_threads=10
```
And the reason is that SETTINGS was parsed by the SELECT parser.
Fix this by push down the SETTINGS from the SELECT to INSERT.

Also note that since INSERT parser does not use ParserQueryWithOutput the
following works:
```
   insert into test_parallel_insert select * from numbers_mt(65535*2) format Null settings max_insert_threads=10
```

(This patch is similar to #12480)

### CODE REVIEW

**Comment by filimonov** [comment]
+1. I saw that problem 'in the wild' several times. It's definitely counterintuitive right now.

File: src/Parsers/InsertQuerySettingsPushDownVisitor.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+class ASTSelectQuery;
+struct SettingChange;
+class SettingsChanges;
+
+/// Pushdown SETTINGS clause to the INSERT from the SELECT query:
+/// (since SETTINGS after SELECT will be parsed by the SELECT parser.)
+///
+/// NOTE: INSERT ... SELECT ... FORMAT Null SETTINGS max_insert_threads=10 works even w/o push down,
+/// since ParserInsertQuery does not ParserInsertQuery.
```
Comment (alexey-milovidov) [inline src/Parsers/InsertQuerySettingsPushDownVisitor.h:17]: Unfinished comment.

**Review by alexey-milovidov — COMMENTED** [review]
LGTM

**Comment by azat** [comment]
>Performance — 2 errors, 2 too long, 10 faster, 78 unstable

```
The test 'parallel_insert' is too slow to run as a whole. Investigate whether the create and fill queries can be sped up
The test 'materialized_view_parallel_insert' is too slow to run as a whole. Investigate whether the create and fill queries can be sped up
```

@akuzm do I understand correctly that the problem is only with the clickhouse-sever from the upstream/master (since queries run in parallel for the patched clickhouse and upstream/master)?
In other words after this PR perf tests should be OK (passed/green) in this particular run.

**Comment by akuzm** [comment]
> > Performance — 2 errors, 2 too long, 10 faster, 78 unstable
> 
> ```
> The test 'parallel_insert' is too slow to run as a whole. Investigate whether the create and fill queries can be sped up
> The test 'materialized_view_parallel_insert' is too slow to run as a whole. Investigate whether the create and fill queries can be sped up
> ```
> 
> @akuzm do I understand correctly that the problem is only with the clickhouse-sever from the upstream/master (since queries run in parallel for the patched clickhouse and upstream/master)?
> In other words after this PR perf tests should be OK (passed/green) in this particular run.

They shouldn't be green because the old server still doesn't read the settings and is too slow, so the error is correct. But you fixed it on the new server, so it's OK to merge. To verify, look at the query times -- they decreased 5x on the new server, this means the setting now works and your fix is correct.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**